exporter/exporterhelper/internal/sized_elements_channel.go

```diff
@@ -0,0 +1,101 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package internal // import "go.opentelemetry.io/collector/exporter/exporterhelper/internal"
+
+import "sync/atomic"
+
+// sizedElementsChannel is a channel wrapper for sized elements with a capacity set to a total size of all the elements.
+// The channel will accept elements until the total size of the elements reaches the capacity.
+type sizedElementsChannel[T any] struct {
+	used *atomic.Uint64
+	cap  uint64
+	ch   chan T
+
+	// The following fields are used by the persistent queue to initialize the queue with the elements recovered from
+	// the disk. Synchronization is not required because the persistent queue doesn't access it concurrently.
+	// dequeue() will decreases the initElementsCount and initTotalSize before the used field.
+	initTotalSize     uint64
+	initElementsCount uint64
+}
+
+// newSizedElementsChannel creates a sized elements channel. Each element is assigned a size by the provided sizer.
+// chanCapacity is the capacity of the underlying channel which usually should be equal to the capacity of the queue to
+// avoid blocking the producer.
+func newSizedElementsChannel[T any](capacity int) *sizedElementsChannel[T] {
+	return &sizedElementsChannel[T]{
+		used: &atomic.Uint64{},
+		cap:  uint64(capacity),
+		ch:   make(chan T, capacity),
+	}
+}
+
+// enqueue puts the element into the queue with the given sized if there is enough capacity.
+// Returns an error if the queue is full. The callback is called before the element is committed to the queue.
+// If the callback returns an error, the element is not put into the queue and the error is returned.
+func (vcq *sizedElementsChannel[T]) enqueue(el T, size uint64, callback func() error) error {
+	if vcq.initTotalSize+vcq.used.Add(size) > vcq.cap {
+		vcq.used.Add(^(size - 1))
+		return ErrQueueIsFull
+	}
+	if callback != nil {
+		if err := callback(); err != nil {
+			vcq.used.Add(^(size - 1))
+			return err
+		}
+	}
+	vcq.ch <- el
+	return nil
+}
+
+// initBulkEnqueue puts the elements into the queue with the given size. It's used by the persistent queue to
+// initialize the queue with the elements recovered from the disk.
+func (vcq *sizedElementsChannel[T]) initBulkEnqueue(els []T, totalSize uint64) {
+	vcq.initElementsCount = uint64(len(els))
+	vcq.initTotalSize = totalSize
```

Comment:
Unnecessary, you can put it into `used` and everything will work correct?

Reply:
The problem is that the total size of the restored queue can be off. So we need to keep it separate and cler this value once the restored queue is drained

Reply:
Why? I am not sure I understand the logic and what you try to achieve. Can you tell me with some numbers for capacity/initial size what you try to achieve? I am not sure, but I still believe to achieve (what I think you try) can be done without these values.

Reply:
At the start of the collector, we restore a number of requests and total items from the disk as separate values. It's guaranteed that the number of requests matches the actual number because of the WAL indexes. But there is no such guarantee for the number of items, it can be off in 2 scenarios:
- User switches from request-based to item-based sizing, so restored number of items is 0.
- We backup the total size of items on disk every 5 reads/writes as an optimization. So, if the collector is killed, the restored number of items can be slightly off. Let's say the collector is killed with 12 requests in the queue (10 items each), but the total number of items was flushed on disk when there were 10 requests in the queue. After restart, we restore 12 requests with 100 items in total instead of 120.

So we need to keep track of the restored queue separately until it's drained.

This is how it's currently implemented in the persistent queue. I'm just moving the logic here. Let me know if you have a better idea.